EDP-380C-28-R2-Lab3.R

```diff
@@ -12,3 +12,358 @@ source("scripts/rmvnorm.R")
+X <- rmvnorm(n, params$mu, params$Sigma)
+
+#-------------------------3.2b------------------------------------------#
+# Method 1: Setting the Total Variance Explained
+set.seed(23921)
+
+# new translate function that gets all params and saves them in a form rmvnorm can deal with and generates the covariance matrix
+translateFun <- function(mu, sigma, rho, betas = NULL, R2 = NULL, muY = NULL, sigmaY = NULL){
+    translated = list(
+        mu = mu, 
+        Sigma = matrix(diag(sigma), length(mu)) %*% 
+            matrix(c(1, rho, rho, 1), nrow = 2) %*% 
+            matrix(diag(sigma), length(mu)),
+        rho = rho, 
+        sigma = sigma,
+        betas = betas,
+        R2 = R2, 
+        muY = muY, 
+        sigmaY = sigmaY
+    )
+}
+
+# translate parameters into form rmvnorm takes
+params <- translateFun(mu = c(5, 10), 
+                       sigma = c(1,2), 
+                       rho = .3, 
+                       betas = c(1, 1), 
+                       R2 = 0.6, 
+                       muY = 10, 
+                       sigmaY = 5)
+
+# matrix way
+varY <- t(params$betas) %*% params$Sigma %*% params$betas # eq 22
+
+# determine variance of error
+varE <- varY * (1/params$R2 - 1) # eq 23
+
+# generate Y
+Y <- params$muY + params$betas[1]*X[,1] + params$betas[2]*X[,2] + rnorm(0, sqrt(varE), n = nrow(X))
+
+# analyze OLS using myLM function
+myMod <- myLM(Y = Y, X = X)
+myMod
+
+summary(lm(Y ~ X[,1] + X[,2]))
+
+# check with lm()
+# summary(lm(Y ~ X[,1] + X[,2]))
+# 
+# Estimate          SE  t value p value
+# Intercept 10.0629400 0.040463429 248.6922       0
+# X1         0.9891621 0.006771543 146.0763       0
+# X2         0.9997359 0.003373688 296.3332       0
+# SD(e)      2.0361731          NA       NA      NA
+# R2         0.7517080          NA       NA      NA
+
+# create a difference bw the population paramters for Y and the results from you generated data
+# intercept, muY
+myMod[1] - params$muY
+# varY, sigmaY
+varY - params$sigmaY
+
+
+#-------------------------3.2c------------------------------------------#
+# Method 2: Setting the Marginal Correlations
+set.seed(123782)
+
+newParams <- list(
+    sigmaX1 = 1, 
+    sigmaX2 = 2, 
+    rhoY1 = .3, 
+    rhoY2 = -.4, 
+    muY = 10, 
+    sigmaY = 5
+)
+
+# corrleation matrix of y, x1, x2
+Rm <- matrix(c(1, newParams$rhoY1, newParams$rhoY2, 
+             newParams$rhoY1, 1, params$rho, 
+             newParams$rhoY2, params$rho, 1), nrow = 3)
+
+# full covariance matrix
+covariancematirx <- diag(c(newParams$sigmaY, 
+                           newParams$sigmaX1, 
+                           newParams$sigmaX2)) %*% 
+                    Rm %*% 
+                    diag(c(newParams$sigmaY,
+                           newParams$sigmaX1,
+                           newParams$sigmaX2))
+
+# solve for betas
+betaZ <- solve(covariancematirx[2:3, 2:3]) %*% covariancematirx[1, 2:3]
```

Comment:
Use `covariancematirx[-1, -1]` this will be more general and robust. Same with the other one.